proto/compiler.bzl

```diff
@@ -32,10 +32,16 @@ load(
 
 GoProtoCompiler = provider()
 
-def go_proto_compile(go, compiler, proto, imports, importpath):
+def go_proto_compile(go, compiler, protos, imports, importpath):
     go_srcs = []
     outpath = None
-    for src in proto.check_deps_sources:
+    dep_sources = depset(direct = [], transitive = [proto.check_deps_sources for proto in protos])
```

Comment:
Since we need to iterate the depset anyway, I think it would be clearer to just have a doubly nested loop, and build `proto_paths` at the same time we're declaring output files.

Constructing a `depset` has a surprising amount of overhead (just in the Java implementation, not anything fundamental), and we should try to avoid it unless either a) we're passing it along a dependency graph and we can avoid iterating it at every step or b) we can pass it to `args.add_all` and avoid iterating it unless the action is executed.

Reply:
We should definitely avoid duplicates, but we're not able to take advantage of `depset` performance here because we're iterating through `check_deps_sources` below, so the cost isn't worth it.

I was thinking a change like this:

```
go_srcs = []
outpath = None
proto_paths = {}
for proto in protos:
    for src in proto.check_deps_sources.to_list():
        path = proto_path(src, proto)
        if path in proto_paths:
            continue
        proto_paths[path] = True
        out = go.declare_file(go, path = importpath + "/" + src.basename[:-len(".proto")], ext = compiler.suffix)
        go_srcs.append(out)
        if outpath == None:
            outpath = out.dirname[:-len(importpath)]
```

WDYT?

The construction of `transitive_descriptor_sets` looks good though. The `depset` makes sense there.

Reply:
This is a valid concern. Let's make it an error.

Ultimately, the `proto_path` is what gets passed to `protoc`, and that's what we use to find the right descriptor to generate code for. We have the full set of descriptors for transitive dependencies, and we don't have another way to identify direct dependencies. It's okay to have the same .proto file repeated (possibly seen through more than one `proto_library`), but it should be an error to have different files with the same effective import path.

So let's do this: `proto_paths` should be a map from `proto_path` values to Bazel `File`s. The code code be changed to something like this:

```
path = proto_path(src, proto)
if path in proto_paths:
    if proto_paths[path] != src:
        fail("proto files {} and {} have the same import path, {}".format(src.path, proto_paths[path].path, path))
    continue
proto_paths[path] = src
```